src/wasm-lib/kcl/src/std/array.rs

```diff
@@ -0,0 +1,96 @@
+use derive_docs::stdlib;
+use schemars::JsonSchema;
+
+use crate::{
+    errors::{KclError, KclErrorDetails},
+    executor::{KclValue, SketchGroup, SourceRange, UserVal},
+    function_param::FunctionParam,
+};
+
+use super::{args::FromArgs, Args, FnAsArg};
+
+/// For each item in an array, update a value.
+pub async fn array_reduce(args: Args) -> Result<KclValue, KclError> {
+    let (array, start, f): (Vec<u64>, SketchGroup, FnAsArg<'_>) = FromArgs::from_args(&args, 0)?;
+    let reduce_fn = FunctionParam {
+        inner: f.func,
+        fn_expr: f.expr,
+        meta: vec![args.source_range.into()],
+        ctx: args.ctx.clone(),
+        memory: *f.memory,
+        dynamic_state: args.dynamic_state.clone(),
+    };
+    inner_array_reduce(array, start, reduce_fn, &args)
+        .await
+        .map(|sg| KclValue::UserVal(UserVal::set(sg.meta.clone(), sg)))
+}
+
+/// Take a starting value. Then, for each element of an array, calculate the next value,
+/// using the previous value and the element.
+/// ```no_run
+/// fn decagon = (radius) => {
+///   let step = (1/10) * tau()
+///   let sketch = startSketchAt([(cos(0)*radius), (sin(0) * radius)])
+///   return arrayReduce([1..10], sketch, (i, sg) => {
+///       let x = cos(step * i) * radius
+///       let y = sin(step * i) * radius
+///       return lineTo([x, y], sg)
+///   })
+/// }
+/// decagon(5.0) |> close(%)
+/// ```
+#[stdlib {
+    name = "arrayReduce",
+}]
+async fn inner_array_reduce<'a>(
+    array: Vec<u64>,
+    start: SketchGroup,
+    reduce_fn: FunctionParam<'a>,
+    args: &'a Args,
+) -> Result<SketchGroup, KclError> {
```

Comment:
Does the return value need to be a SketchGroup? Can it be a KclValue?

Reply:
Right now, yes it does, seem PR description :( hopefully we can relax that.

Reply:
Ah, sorry for not reading.